### ISSUE TITLE
linspace returns wrong result for int32

### ISSUE THREAD

**Issue by mthrok** [pytorch/pytorch#40118]
## 🐛 Bug

`linspace` with `int32` dtype returns wrong results when input values are big. This happens on nightly build (`macOS`) and source build (`ubuntu`).

## To Reproduce

```
$ python -c 'import torch;print(torch.linspace(-2147483647, 2147483647, 10, dtype=torch.int32))'
tensor([-2147483647, -2147483647, -2147483647, -2147483647, -2147483647,
         2147483647,  2147483647,  2147483647,  2147483647,  2147483647],
       dtype=torch.int32)
```

```
$ python -c 'import torch;print(torch.linspace(-2000000000, 2147483647, 10, dtype=torch.int32))'
tensor([-2000000000, -2016387072, -2032774144, -2049161216, -2065548288,
        -2147483648, -2147483648, -2147483648, -2147483648,  2147483647],
       dtype=torch.int32)
```

```
$ python -c 'import torch;print(torch.linspace(-1000000000, 2147483647, 10, dtype=torch.int32))'
tensor([-1000000000, -1127498183, -1254996366, -1382494549, -1509992732,
        -2147483648, -2147483648, -2147483648, -2147483648,  2147483647],
       dtype=torch.int32)
```

## Expected behavior

linear interpolation is correct

## Environment

<details><summary><code>Mac</code></summary>

```
Collecting environment information...
PyTorch version: 1.6.0.dev20200616
Is debug build: No
CUDA used to build PyTorch: None

OS: Mac OSX 10.15.5
GCC version: Could not collect
CMake version: Could not collect

Python version: 3.7
Is CUDA available: No
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.6.0.dev20200616
[conda] blas                      1.0                         mkl
[conda] mkl                       2019.4                      233
[conda] mkl-service               2.3.0            py37hfbe908c_0
[conda] mkl_fft                   1.0.15           py37h5e564d8_0
[conda] mkl_random                1.1.0            py37ha771720_0
[conda] numpy                     1.18.1           py37h7241aed_0
[conda] numpy-base                1.18.1           py37h6575580_1
[conda] pytorch                   1.6.0.dev20200616         py3.7_0    pytorch-nightly
```
</details>

<details><summary><code>Ubuntu</code></summary>

```
Collecting environment information...
PyTorch version: 1.6.0a0+ec1833b
Is debug build: No
CUDA used to build PyTorch: 10.1

OS: Ubuntu 18.04.3 LTS
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
CMake version: version 3.10.2

Python version: 3.8
Is CUDA available: Yes
CUDA runtime version: 10.1.243
GPU models and configuration:
GPU 0: Quadro GP100
GPU 1: Quadro GP100

Nvidia driver version: 418.116.00
cuDNN version: /usr/lib/x86_64-linux-gnu/libcudnn.so.7.6.5

Versions of relevant libraries:
[pip] numpy==1.18.1
[pip] torch==1.6.0a0+ec1833b
[pip] torchaudio==0.6.0a0+2fa6cc7
[pip] torchtext==0.6.0a0+e823aca
[pip] torchvision==0.7.0a0+f9ef235
[conda] magma-cuda101             2.5.2                         1    pytorch
[conda] mkl                       2020.1                      217
[conda] mkl-include               2020.1                      217    conda-forge
[conda] numpy                     1.18.1           py38h8854b6b_1    conda-forge
[conda] torch                     1.6.0a0+ec1833b           dev_0    <develop>
[conda] torchaudio                0.6.0a0+2fa6cc7           dev_0    <develop>
[conda] torchtext                 0.6.0a0+e823aca           dev_0    <develop>
[conda] torchvision               0.7.0a0+f9ef235           dev_0    <develop>
```
</details>

**Comment by malfet** [comment]
Problem exist in 1.5.0 as well and seem to stem from the fact that linspace range is greater than int32

**Comment by mthrok** [comment]
> linspace range is greater than int32

@malfet What do you mean? Isn't `[-2147483648, 2147483647]` valid range of `int32`??

**Comment by malfet** [comment]
@mthrok have a look at this one:
https://github.com/pytorch/pytorch/blob/3d8de74e176b2ab39cbf0adc2bdbe229d14bccfa/aten/src/ATen/native/cpu/RangeFactoriesKernel.cpp#L21

`2147483647 - (-2147483648) == 0xffffffff`, which is -1 when cast back to `int32`

**Comment by mthrok** [comment]
@malfet Oh I see. Thanks for the explanation.

**Comment by malfet** [comment]
Something like that should fix it: https://github.com/pytorch/pytorch/pull/40132 though it would have negative effect on the accuracy for some  large int64 values.
